File: internal/images/contexts/_shared/models/constraints.py
"""Constraints configuration schema."""

from typing import Literal

from pydantic import BaseModel, ConfigDict, Field

from agency_core.models.egress import EgressMode


class HardLimit(BaseModel):
    model_config = ConfigDict(extra="forbid")
    rule: str
    reason: str


class EscalationConfig(BaseModel):
    model_config = ConfigDict(extra="forbid")
    always_escalate: list[str] = []
    flag_before_proceeding: list[str] = []


class NotifyConfig(BaseModel):
    """Notification targets for budget alerts."""

    model_config = ConfigDict(extra="forbid")

    webhook: str | None = None
    email: str | None = None
    log: bool = True


class BudgetConfig(BaseModel):
    """Per-agent LLM spend budget controls."""

    model_config = ConfigDict(extra="forbid")

    mode: Literal["hard", "soft", "notify"] = "notify"
    soft_limit: float = Field(default=0.0, ge=0.0)  # dollars, alert threshold
    hard_limit: float = Field(default=0.0, ge=0.0)   # dollars, block threshold
    max_daily_usd: float = Field(default=0.0, ge=0.0)       # 0 = unlimited
    max_session_usd: float = Field(default=0.0, ge=0.0)
    max_total_usd: float = Field(default=0.0, ge=0.0)
    warning_threshold_pct: int = Field(default=80, ge=1, le=100)
    notify: NotifyConfig = NotifyConfig()


class MCPPolicy(BaseModel):
    """Per-agent MCP tool access policy.

    Controls which MCP servers and individual tools the agent can use.
    Default mode is allowlist — only explicitly allowed servers/tools
    are available. Set mode to "denylist" to allow all except denied.
    """

    model_config = ConfigDict(extra="forbid")

    mode: Literal["allowlist", "denylist"] = "denylist"
    allowed_servers: list[str] = []
    denied_servers: list[str] = []
    allowed_tools: list[str] = []
    denied_tools: list[str] = []
    pinned_hashes: dict[str, str] = {}  # server_name -> sha256 hex of command binary

    def is_server_allowed(self, server_name: str) -> bool:
        """Check if an MCP server is allowed by this policy."""
        if self.mode == "allowlist":
            return server_name in self.allowed_servers
        else:
            return server_name not in self.denied_servers

    def is_tool_allowed(self, tool_name: str) -> bool:
        """Check if an individual MCP tool is allowed by this policy."""
        if self.denied_tools and tool_name in self.denied_tools:
            return False
        if self.allowed_tools:
            return tool_name in self.allowed_tools
        return True


class NetworkConfig(BaseModel):
    """Network access configuration."""

    model_config = ConfigDict(extra="forbid")

    egress_mode: EgressMode = "denylist"


class IdentityConstraint(BaseModel):
    model_config = ConfigDict(extra="forbid")
    role: str
    purpose: str


class ConstraintsConfig(BaseModel):
    """Schema for constraints.yaml."""

    # Use "ignore" so that legacy constraints.yaml files containing an
    # "autonomy" key (or any other unknown keys) are silently accepted.
    model_config = ConfigDict(extra="ignore")

    version: str = "0.1"
    agent: str
    identity: IdentityConstraint
    hard_limits: list[HardLimit] = []
    escalation: EscalationConfig = EscalationConfig()
    network: NetworkConfig = NetworkConfig()
    budget: BudgetConfig = BudgetConfig()
    mcp: MCPPolicy = MCPPolicy()

File: internal/models/constraints.go
// agency-gateway/internal/models/constraints.go
package models

// HardLimit is a named rule that the agent must never violate.
// Also used by preset models (preset.go will reference this type).
type HardLimit struct {
	Rule   string `yaml:"rule" validate:"required"`
	Reason string `yaml:"reason" validate:"required"`
}

// EscalationConfig controls when the agent must escalate to a human.
type EscalationConfig struct {
	AlwaysEscalate        []string `yaml:"always_escalate"`
	FlagBeforeProceeding  []string `yaml:"flag_before_proceeding"`
}

// NotifyConfig specifies notification targets for budget events.
type NotifyConfig struct {
	Webhook string `yaml:"webhook"`
	Email   string `yaml:"email"`
	Log     bool   `yaml:"log" default:"true"`
}

// BudgetConfig controls cost budget enforcement.
type BudgetConfig struct {
	Mode                string      `yaml:"mode" validate:"oneof=hard soft notify" default:"notify"`
	SoftLimit           float64     `yaml:"soft_limit" validate:"gte=0"`
	HardLimit           float64     `yaml:"hard_limit" validate:"gte=0"`
	MaxDailyUSD         float64     `yaml:"max_daily_usd" validate:"gte=0"`
	MaxSessionUSD       float64     `yaml:"max_session_usd" validate:"gte=0"`
	MaxTotalUSD         float64     `yaml:"max_total_usd" validate:"gte=0"`
	WarningThresholdPct int         `yaml:"warning_threshold_pct" validate:"gte=1,lte=100" default:"80"`
	Notify              NotifyConfig `yaml:"notify"`
}

// MCPPolicy controls which MCP servers and tools the agent may use.
type MCPPolicy struct {
	Mode           string            `yaml:"mode" validate:"oneof=allowlist denylist" default:"denylist"`
	AllowedServers []string          `yaml:"allowed_servers"`
	DeniedServers  []string          `yaml:"denied_servers"`
	AllowedTools   []string          `yaml:"allowed_tools"`
	DeniedTools    []string          `yaml:"denied_tools"`
	PinnedHashes   map[string]string `yaml:"pinned_hashes"`
}

// IsServerAllowed returns true if the named server is permitted by this policy.
func (p *MCPPolicy) IsServerAllowed(serverName string) bool {
	if p.Mode == "allowlist" {
		for _, s := range p.AllowedServers {
			if s == serverName {
				return true
			}
		}
		return false
	}
	// denylist mode
	for _, s := range p.DeniedServers {
		if s == serverName {
			return false
		}
	}
	return true
}

// IsToolAllowed returns true if the named tool is permitted by this policy.
func (p *MCPPolicy) IsToolAllowed(toolName string) bool {
	for _, t := range p.DeniedTools {
		if t == toolName {
			return false
		}
	}
	if len(p.AllowedTools) > 0 {
		for _, t := range p.AllowedTools {
			if t == toolName {
				return true
			}
		}
		return false
	}
	return true
}

// NetworkConfig controls network egress behaviour for the agent.
type NetworkConfig struct {
	EgressMode string `yaml:"egress_mode" validate:"oneof=denylist allowlist supervised-strict supervised-permissive" default:"denylist"`
}

// IdentityConstraint records the agent's declared role and purpose.
type IdentityConstraint struct {
	Role    string `yaml:"role" validate:"required"`
	Purpose string `yaml:"purpose" validate:"required"`
}

// ConstraintsConfig is the schema for constraints.yaml.
type ConstraintsConfig struct {
	Version    string             `yaml:"version" default:"0.1"`
	Agent      string             `yaml:"agent" validate:"required"`
	Identity   IdentityConstraint `yaml:"identity" validate:"required"`
	HardLimits []HardLimit        `yaml:"hard_limits"`
	Escalation EscalationConfig   `yaml:"escalation"`
	Network    NetworkConfig      `yaml:"network"`
	Budget     BudgetConfig       `yaml:"budget"`
	MCP        MCPPolicy          `yaml:"mcp"`
}
